### DISCUSSION TITLE
Reset wizard?

### DISCUSSION THREAD

**Discussion by Indemnity83** [spatie/laravel-livewire-wizard#69] [Q&A] [answered]
Is there a way to reset the wizard? At the end of the wizard my action is completed but short of a full page reload I don't see how to set the wizard back to the initial state and on the first step so it's ready to be called again.

Am I just missing something?

**Comment by ju5t** [comment]
The `showStep` method in a step just emits an event.

```php
$this->emitUp('showStep', $stepName, $this->state()->currentStep());
```

I think if you pass an empty array it would just reset all state.

**Comment by Indemnity83** [comment, marked as answer]
@ju5t that didn't quite work, but got me on the right path to a solution. 

For anybody who stumbles on this, the `mountMountsWizard` function inside the `MountsWizard` trait is what is responsible for setting up the steps and states. so I can use it to reset the state.

So I added a listener for a reset event that calls the `mountMountsWizard()` function. 
```php
class MyWizardComponent extends WizardComponent
{
    // This is ugly... have to override the $listeners to add our listener
    protected $listeners = [
        'previousStep',
        'nextStep',
        'resetWizard',
        'showStep',
    ];

    public function resetWizard()
    {
        // Get the first step from the collection
        $firstStep = $this->stepNames()->first();

        // Create an array of step names with empty states
        $initialState = $this->stepNames()->mapWithKeys(function ($step) {
            return [$step => []];
        })->all();

        $this->mountMountsWizard($firstStep, $initialState);
    }

    // ... rest of class
}
```

**Reply by ju5t** [reply]
Great!

This would be an alternative for the listeners.

```php
protected function getListeners(): array
{
    return [...parent::getListeners(), 'resetWizard'];
}
```

If your PHP version doesn't support the splat operator you can use `array_merge` instead.

**Comment by Kotzilla** [comment]
Hi, I have question
How can i call reset from StepComponent ?

**Comment by mauriciodinamarca** [comment]
Thanks! Worked for me, just needed to update using livewire v3:
From the StepComponent call 
`$this->dispatch('resetWizard');`
And the Custom Wizard Component update as required (recommended by @ju5t and @Indemnity83) :

```php
use Livewire\Attributes\On;
use Spatie\LivewireWizard\Components\WizardComponent;

abstract class CustomWizardComponent extends WizardComponent
{
    protected function getListeners(): array
    {
        return [...parent::getListeners(), 'resetWizard'];
    }

    #[On('resetWizard')]
    public function resetWizard()
    {
        // Get the first step from the collection
        $firstStep = $this->stepNames()->first();

        // Create an array of step names with empty states
        $initialState = $this->stepNames()->mapWithKeys(function ($step) {
            return [$step => []];
        })->all();

        $this->mountMountsWizard($firstStep, $initialState);
    }
}
```
